homeassistant/components/mobile_app/__init__.py

```diff
@@ -0,0 +1,326 @@
+
+    webhook_type = req_data[ATTR_WEBHOOK_TYPE]
+
+    webhook_payload = req_data.get(ATTR_WEBHOOK_DATA, {})
+
+    if req_data[ATTR_WEBHOOK_ENCRYPTED]:
+        enc_data = req_data[ATTR_WEBHOOK_ENCRYPTED_DATA]
+        webhook_payload = _decrypt_payload(device[CONF_SECRET], enc_data)
+
+    try:
+        data = WEBHOOK_SCHEMAS[webhook_type](webhook_payload)
+    except vol.Invalid as ex:
+        return json_response(vol.humanize.humanize_error(webhook_payload, ex),
+                             status=HTTP_BAD_REQUEST)
+
+    if webhook_type == WEBHOOK_TYPE_CALL_SERVICE:
+
+        with AsyncTrackStates(hass) as changed_states:
+            try:
+                await hass.services.async_call(data[ATTR_DOMAIN],
+                                               data[ATTR_SERVICE],
+                                               data[ATTR_SERVICE_DATA],
+                                               True,
+                                               context(request))
+            except (vol.Invalid, ServiceNotFound):
+                raise HTTPBadRequest()
+
+        return json_response(changed_states)
+    elif webhook_type == WEBHOOK_TYPE_FIRE_EVENT:
+        event_type = data[ATTR_EVENT_TYPE]
+        hass.bus.fire(event_type, data[ATTR_EVENT_DATA], ha.EventOrigin.remote,
+                      context(request))
+        return json_response({"message": "Event {} fired.".format(event_type)})
+    elif webhook_type == WEBHOOK_TYPE_RENDER_TEMPLATE:
+        try:
+            tpl = template.Template(data[ATTR_TEMPLATE], hass)
+            rendered = tpl.async_render(data.get(ATTR_TEMPLATE_VARIABLES))
+            return json_response({"rendered": rendered})
+        except (ValueError, TemplateError) as ex:
+            return json_response("Error rendering template: {}".format(ex),
+                                 status=HTTP_BAD_REQUEST)
+
+    elif webhook_type == WEBHOOK_TYPE_UPDATE_LOCATION:
+        await hass.services.async_call(DEVICE_TRACKER_DOMAIN,
+                                       DEVICE_TRACKER_SEE, data)
+        return json_response([])
+
+
+def supports_encryption():
+    """Test if we support encryption."""
+    try:
+        import nacl   # noqa pylint: disable=unused-import
+        return True
+    except OSError:
+        return False
+
+
+def device_for_webhook_id(hass, webhook_id):
+    """Return the device name for the webhook ID."""
+    for device_name, device in hass.data[DOMAIN].items():
+        if device.get(CONF_WEBHOOK_ID) == webhook_id:
+            return device
+    return None
+
+
+async def async_setup(hass, config):
+    """Set up the mobile app component."""
+    conf = config.get(DOMAIN)
+
+    store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+    app_config = await store.async_load()
+    if app_config is None:
+        app_config = {}
+
+    hass.data[DOMAIN] = app_config
+
+    for name, device in app_config.items():
+        if CONF_WEBHOOK_ID in device:
```

Comment:
Can it be that there is no webhook_id ?

Reply:
If there is no webhook, the app can't communicate back, so there is no use for it?